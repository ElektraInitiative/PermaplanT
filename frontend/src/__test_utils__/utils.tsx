import { onError } from '@/config/react_query';
import { QueryCache, QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { FormProvider, useForm } from 'react-hook-form';
import { ToastContainer } from 'react-toastify';

export interface TestDto {
  foo: string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const FormWrapper = (props: any) => {
  const formMethods = useForm<TestDto>();

  return <FormProvider {...formMethods}>{props.children}</FormProvider>;
};

export const createQueryHookWrapper = () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
    queryCache: new QueryCache({
      onError: onError,
    }),
  });

  return function Wrapper({ children }: { children: React.ReactNode }) {
    return (
      <QueryClientProvider client={queryClient}>
        {children}
        <ToastContainer></ToastContainer>
      </QueryClientProvider>
    );
  };
};
