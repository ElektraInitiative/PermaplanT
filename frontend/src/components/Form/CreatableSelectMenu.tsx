import filterObject from '../../utils/filterObject';
import { SelectOption } from './SelectMenuTypes';
import { useState } from 'react';
import { Control, Controller, FieldValues, Path } from 'react-hook-form';
import {
  ActionMeta,
  ClassNamesConfig,
  GroupBase,
  MultiValue,
  SingleValue,
  StylesConfig,
} from 'react-select';
import CreatableSelect from 'react-select/creatable';

export interface CreatableSelectMenuProps<
  T extends FieldValues,
  Option = SelectOption,
  IsMulti extends boolean = false,
> {
  /** When this boolean flag is set to true, the user may select multiple options at once (default false).                                                                          */
  isMulti?: IsMulti;
  /** Unique name of this component.                                                                                                                                                */
  id: Path<T>;
  /** Short description that will be displayed above the input field if set.                                                                                                        */
  labelText?: string;
  /** React hook form control (See https://www.react-hook-form.com/api/useform/control/ for further information).                                                                   */
  control?: Control<T, unknown>;
  /** Options that can be selected by the user. I.e. the contents of this component.                                                                                                */
  options: Option[];
  /** If set to true, the user has to select an option before the form can be completed (default false).                                                                            */
  required?: boolean;
  /** Text that is displayed instead of the input if it has not been selected yet.                                                                                                  */
  placeholder?: string;
  /** Callback that is invoked every time the user selects a new option. The single argument represents the selected option.                                                        */
  handleOptionsChange?: (
    option: SingleValue<Option> | MultiValue<Option>,
    actionMeta: ActionMeta<Option>,
  ) => void;
  /** Gets called if a new item was created with the users input as the only argument. If not set, the component will automatically add all created items to the available options. */
  handleCreate?: (inputValue: string) => void;
  /** Callback that is invoked if the user performs ANY input. This includes type and click events among others.                                                                    */
  onChange?: () => void;
  /** Callback that is invoked if the fields content changes. */
  onInputChange?: (inputValue: string) => void;
}

/**
 * Similar to SelectMenu, but allows the user to create new menu entries.
 * @param props.isMulti: When this boolean flag is set to true, the user may select multiple options at once (default false).
 * @param props.id: Unique name of this component.
 * @param props.labelText: Short description that will be displayed above the input field if set.
 * @param props.control: React hook form control (See https://www.react-hook-form.com/api/useform/control/ for further information).
 * @param props.options: Options that can be selected by the user. I.e. the contents of this component.
 * @param props.required: If set to true, the user has to select an option before the form can be completed (default false).
 * @param props.placeholder: Text that is displayed instead of the input if it has not been selected yet.
 * @param props.handleOptionsChange: Callback that is invoked every time the user selects a new option. The single argument represents the selected option.
 * @param props.handleCreate: Gets called if a new item was created with the users input as the only argument. If not set, the component will automatically add all created items to the available options.
 * @param props.onChange: Callback that is invoked if the user performs ANY input. This includes type and click events among others.
 * @param props.onInputChange: Callback that is invoked if the fields content changes.
 */
export default function CreatableSelectMenu<
  T extends FieldValues,
  Option = SelectOption,
  IsMulti extends boolean = false,
>({
  isMulti = false as IsMulti,
  id,
  labelText,
  control,
  options,
  required = false,
  placeholder,
  handleOptionsChange,
  handleCreate,
  onChange,
  onInputChange,
}: CreatableSelectMenuProps<T, Option, IsMulti>) {
  const customClassNames: ClassNamesConfig<Option, IsMulti, GroupBase<Option>> = {
    menu: () => 'bg-neutral-100 dark:bg-neutral-50-dark',
    control: (state) => {
      return `
        h-[44px] bg-neutral-200 rounded border
        dark:bg-neutral-50-dark focus:border-primary-500
        hover:border-primary-500 dark:focus:border-primary-300 dark:hover:border-primary-300
        ${
          state.isFocused
            ? ' border-primary-500 dark:border-primary-300'
            : ' dark:border-neutral-400-dark border-neutral-500'
        }
      `;
    },
    option: (state) => {
      return `
        hover:bg-neutral-200 dark:hover:bg-neutral-400-dark
        ${state.isFocused ? ' bg-neutral-300 dark:bg-neutral-500' : ''}
        ${state.isSelected ? ' bg-primary-500' : ''}
      `;
    },
    valueContainer: () => 'flex-nowrap',
    multiValue: () => 'bg-neutral-400 dark:bg-neutral-400-dark',
    multiValueRemove: () => 'hover:bg-neutral-500',
  };
  const customStyles: StylesConfig<Option, IsMulti, GroupBase<Option>> = {
    // remove css attributes from predefined styles
    // this needs to be done so the custom css classes take effect
    control: (styles) =>
      filterObject(styles, [
        'border',
        'borderColor',
        'borderRadius',
        'boxShadow',
        'color',
        '&:hover',
      ]),
    option: (styles) => filterObject(styles, ['backgroundColor', 'color']),
    singleValue: (styles) => filterObject(styles, ['color']),
    multiValue: (styles) => filterObject(styles, ['color']),
    multiValueLabel: (styles) => filterObject(styles, ['color']),
    multiValueRemove: (styles) => filterObject(styles, ['color']),
  };

  const [inputValue, setInputValue] = useState('');

  return (
    <div>
      {labelText && (
        <label htmlFor={id} className="mb-2 block text-sm font-medium text-white">
          {labelText}
          {required ? <span className="text-red-800"> *</span> : <></>}
        </label>
      )}

      <Controller
        name={id}
        control={control}
        render={() => (
          <CreatableSelect
            name={id}
            isClearable
            onChange={handleOptionsChange}
            onCreateOption={handleCreate}
            inputValue={inputValue}
            placeholder={placeholder}
            options={options}
            isMulti={isMulti}
            styles={customStyles}
            classNames={customClassNames}
            required={required}
            onInputChange={(value, event) => {
              // prevent the text from disapearing when clicking inside the input field
              if (event.action === 'input-change' || event.action === 'set-value') {
                setInputValue(value);
              }
              onChange?.();
              onInputChange?.(value);
            }}
          />
        )}
      />
    </div>
  );
}
