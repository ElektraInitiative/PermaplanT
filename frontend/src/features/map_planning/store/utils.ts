import type { GetFn } from './MapStoreTypes';

/**
 * If any of the selected nodes can not be found on the map, clear the selection.
 */
export function clearInvalidSelection(get: GetFn) {
  const stage = get().stageRef.current;
  const transformer = get().transformer.current;
  const selectedNodeIds = transformer?.getNodes().map((n) => n.id()) || [];

  const selectedNodesAreVisible = selectedNodeIds.every((id) => stage?.findOne(`#${id}`));

  if (!selectedNodesAreVisible) {
    transformer?.nodes([]);
    get().selectPlantings(null);
  }
}
