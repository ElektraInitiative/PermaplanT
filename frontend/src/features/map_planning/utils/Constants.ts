export const MAP_PIXELS_PER_METER = 100;
export const TEN_CENTIMETERS = 10;
export const ONE_METER = 100;
