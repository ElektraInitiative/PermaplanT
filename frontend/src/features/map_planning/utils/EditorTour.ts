import { TFunction } from 'i18next';
import { ShepherdOptionsWithType, Tour } from 'react-shepherd';

export function getSteps(t: TFunction): ShepherdOptionsWithType[] {
  t('ok');
  return [
    {
      id: 'welcome',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'toolboxIntro',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'layersIntro',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'timelineIntro',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'szenarioIntro',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'baseLayerIntro',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'baseLayerToolbar',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'baseImageSelect',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'plantsLayerIntro',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'plantsLayerSelect',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'plantsLayerRightToolbar',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'plantsLayerLeftToolbar',
      buttons: [
        {
          text: 'Next',
          type: 'next',
        },
      ],
    },
    {
      id: 'finish',
      buttons: [
        {
          text: 'Finish',
          type: 'next',
        },
      ],
    },
  ];
}

export const tourOptions: Tour.TourOptions = {
  defaultStepOptions: {
    cancelIcon: {
      enabled: true,
    },
  },
  useModalOverlay: true,
};
