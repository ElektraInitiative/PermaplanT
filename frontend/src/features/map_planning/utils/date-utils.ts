/**
 * Takes a string in the format of "YYYY-MM-DD" and converts it to a Date object.
 */
export function convertToDate(date: string) {
  const [year, month, day] = date.split('-');
  return new Date(parseInt(year), parseInt(month) - 1, parseInt(day));
}

/**
 * Takes a Date object and converts it to a string in the format of "YYYY-MM-DD".
 */
export function convertToDateString(date: Date) {
  const year = date.getFullYear();
  const month = (date.getMonth() + 1).toString().padStart(2, '0');
  const day = date.getDate().toString().padStart(2, '0');

  return `${year}-${month}-${day}`;
}

/**
 * Returns the short name of a month in the given language.
 *
 * @param monthNumber month from 1 to 12
 * @param language language code, e.g. 'en-US'
 * @returns short name of month in given language (e.g. 'Jan')
 */

export function getShortMonthNameFromNumber(monthNumber: number, language?: string) {
  const date = new Date();
  date.setMonth(monthNumber - 1);

  return date.toLocaleString(language, { month: 'short' });
}
