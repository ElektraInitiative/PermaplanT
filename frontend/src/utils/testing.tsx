import { FormProvider, useForm } from 'react-hook-form';

export interface TestDto {
  foo: string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const FormWrapper = (props: any) => {
  const formMethods = useForm<TestDto>();

  return <FormProvider {...formMethods}>{props.children}</FormProvider>;
};
