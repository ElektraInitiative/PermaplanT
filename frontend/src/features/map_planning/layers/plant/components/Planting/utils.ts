import { KonvaEventObject } from 'konva/lib/Node';
import { PlantsSummaryDto } from '@/api_types/definitions';
import {
  setTooltipPositionToMouseCursor,
  showTooltipWithContent,
} from '@/features/map_planning/utils/Tooltip';
import { getNameFromPlant, getPlantNameFromAdditionalNameAndPlant } from '@/utils/plant-naming';

export function triggerPlantSelectionInGuidedTour(): void {
  const placeEvent = new Event('selectPlant');
  document.getElementById('canvas')?.dispatchEvent(placeEvent);
}

export function isUsingModifierKey(e: KonvaEventObject<MouseEvent>): boolean {
  return e.evt.ctrlKey || e.evt.shiftKey || e.evt.metaKey;
}

export function placeTooltip(
  plant: PlantsSummaryDto | undefined,
  additionalName: string | undefined,
) {
  if (!plant) return;

  setTooltipPositionToMouseCursor();
  if (!additionalName) {
    showTooltipWithContent(getNameFromPlant(plant));
  } else {
    showTooltipWithContent(getPlantNameFromAdditionalNameAndPlant(additionalName, plant));
  }
}
