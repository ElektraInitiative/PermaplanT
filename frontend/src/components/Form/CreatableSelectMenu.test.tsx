import { FormWrapper } from '../../utils/testing';
import CreatableSelectMenu from './CreatableSelectMenu';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <FormWrapper>
          <CreatableSelectMenu id="test" labelText="Some Label" options={[]} />
        </FormWrapper>
      </MemoryRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('calls onInputChange on input change', async () => {
  const callback = jest.fn();

  const { getByRole } = render(
    <MemoryRouter>
      <FormWrapper>
        <CreatableSelectMenu
          id="test"
          labelText="Test"
          placeholder="Test"
          options={[]}
          onInputChange={callback}
        />
      </FormWrapper>
    </MemoryRouter>,
  );

  await userEvent.click(getByRole('combobox'));
  await userEvent.paste('Hello World!');

  expect(callback).toBeCalled();
  expect(callback).toBeCalledWith('Hello World!');
});

it('calls onChange on input change', async () => {
  const callback = jest.fn();
  const { getByRole } = render(
    <MemoryRouter>
      <FormWrapper>
        <CreatableSelectMenu
          id="test"
          labelText="Test"
          placeholder="Test"
          options={[]}
          onChange={callback}
        />
      </FormWrapper>
    </MemoryRouter>,
  );

  await userEvent.click(getByRole('combobox'));
  await userEvent.paste('Hello World!');

  expect(callback).toBeCalled();
});

it('calls handleOptionsChange on options change', async () => {
  const callback = jest.fn();
  const { getByRole, getByText } = render(
    <MemoryRouter>
      <FormWrapper>
        <CreatableSelectMenu
          id="test"
          labelText="Test"
          placeholder="Test"
          options={[{ value: 'foo', label: 'foo' }]}
          handleOptionsChange={callback}
        />
      </FormWrapper>
    </MemoryRouter>,
  );

  await userEvent.click(getByRole('combobox'));
  await userEvent.paste('fo');
  await userEvent.click(getByText('foo'));

  expect(callback).toBeCalled();
});

it('creates new option if none exists', async () => {
  const callback = jest.fn();
  const { getByRole } = render(
    <MemoryRouter>
      <FormWrapper>
        <CreatableSelectMenu
          id="test"
          labelText="Test"
          placeholder="Test"
          options={[{ value: 'foo', label: 'foo' }]}
          handleOptionsChange={callback}
        />
      </FormWrapper>
    </MemoryRouter>,
  );

  await userEvent.click(getByRole('combobox'));
  await userEvent.keyboard('does not exist{Enter}');

  const first_arg = {
    __isNew__: true,
    label: 'does not exist',
    value: 'does not exist',
  };
  const second_arg = {
    action: 'create-option',
    name: 'test',
    option: {
      __isNew__: true,
      label: 'does not exist',
      value: 'does not exist',
    },
  };
  expect(callback).toBeCalledWith(first_arg, second_arg);
});
