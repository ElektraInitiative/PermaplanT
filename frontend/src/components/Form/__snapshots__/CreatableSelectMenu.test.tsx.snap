// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
<div>
  <label
    className="mb-2 block text-sm font-medium text-white"
    htmlFor="test"
  >
    Some Label
  </label>
  <div
    className=" css-b62m3t-container"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-2-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
      role="log"
    />
    <div
      className="h-[44px] bg-neutral-200 rounded border
        dark:bg-neutral-50-dark focus:border-primary-500
        hover:border-primary-500 dark:focus:border-primary-300 dark:hover:border-primary-300
         dark:border-neutral-400-dark border-neutral-500 css-qz5xir-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="flex-nowrap css-1fdsijx-ValueContainer"
      >
        <div
          className=" css-1jqq78o-placeholder"
          id="react-select-2-placeholder"
        >
          Select...
        </div>
        <div
          className=" css-qbdosj-Input"
          data-value=""
        >
          <input
            aria-activedescendant=""
            aria-autocomplete="list"
            aria-describedby="react-select-2-placeholder"
            aria-expanded={false}
            aria-haspopup={true}
            aria-required={false}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className=""
            disabled={false}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className=" css-1hb7zxy-IndicatorsContainer"
      >
        <span
          className=" css-1u9des2-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className=" css-1xc3v61-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-tj5bde-Svg"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
    <input
      name="test"
      type="hidden"
      value=""
    />
  </div>
</div>
`;
